<a href="https://colab.research.google.com/github/Nkuluh/Market/blob/main/Employment_outside_formal_sector.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [71]:
pip install mylib

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for mylib: filename=mylib-0.0.1-py3-none-any.whl size=1006 sha256=1e0f80c5e0620ca5f72be24baecff1b46b2b135ea946c4a5ad2a7c869a7fa5a3
  Stored in directory: /root/.cache/pip/wheels/c7/f4/8e/7ce382f7da1725868d2b769b72c419136bd259535df13a1fc4
Successfully built mylib


In [7]:
library(ggplot2)

NameError: ignored

#Importing Useful libraries

In [57]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

#Reading the dataset as a

In [8]:
a= pd.read_csv('EMP.csv')

#View the data to see the number of rows and colums

In [9]:
a

ref_area.label                                    indicator.label  \
0        Afghanistan  Employment outside the formal sector by sex an...   
1        Afghanistan  Employment outside the formal sector by sex an...   
2        Afghanistan  Employment outside the formal sector by sex an...   
3        Afghanistan  Employment outside the formal sector by sex an...   
4        Afghanistan  Employment outside the formal sector by sex an...   
...              ...                                                ...   
32022       Zimbabwe  Employment outside the formal sector by sex an...   
32023       Zimbabwe  Employment outside the formal sector by sex an...   
32024       Zimbabwe  Employment outside the formal sector by sex an...   
32025       Zimbabwe  Employment outside the formal sector by sex an...   
32026       Zimbabwe  Employment outside the formal sector by sex an...   

                    source.label    sex.label  \
0      LFS - Labour Force Survey   Sex: Total   
1      LFS - Labour Force Survey   Sex: Total   
2      LFS - Labour Force Survey   Sex: Total   
3      LFS - Labour Force Survey    Sex: Male   
4      LFS - Labour Force Survey    Sex: Male   
...                          ...          ...   
32022  LFS - Labour Force Survey  Sex: Female   
32023  LFS - Labour Force Survey  Sex: Female   
32024  LFS - Labour Force Survey  Sex: Female   
32025  LFS - Labour Force Survey  Sex: Female   
32026  LFS - Labour Force Survey  Sex: Female   

                                          classif1.label  time  obs_value  \
0      Economic activity (Agriculture, Non-Agricultur...  2021   6059.359   
1      Economic activity (Agriculture, Non-Agricultur...  2021   1884.001   
2      Economic activity (Agriculture, Non-Agricultur...  2021   4175.358   
3      Economic activity (Agriculture, Non-Agricultur...  2021   4325.320   
4      Economic activity (Agriculture, Non-Agricultur...  2021   1237.423   
...                                                  ...   ...        ...   
32022  Economic activity (ISIC-Rev.4): R. Arts, enter...  2019        NaN   
32023  Economic activity (ISIC-Rev.4): S. Other servi...  2019     27.658   
32024  Economic activity (ISIC-Rev.4): T. Activities ...  2019    119.138   
32025  Economic activity (ISIC-Rev.4): U. Activities ...  2019        NaN   
32026  Economic activity (ISIC-Rev.4): X. Not elsewhe...  2019        NaN   

      obs_status.label note_classif.label  \
0                  NaN                NaN   
1                  NaN                NaN   
2                  NaN                NaN   
3                  NaN                NaN   
4                  NaN                NaN   
...                ...                ...   
32022       Unreliable                NaN   
32023  Break in series                NaN   
32024  Break in series                NaN   
32025       Unreliable                NaN   
32026       Unreliable                NaN   

                       note_indicator.label  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
...                                     ...   
32022  Break in series: Methodology revised   
32023  Break in series: Methodology revised   
32024  Break in series: Methodology revised   
32025  Break in series: Methodology revised   
32026  Break in series: Methodology revised   

                                       note_source.label  
0      Repository: ILO-STATISTICS - Micro data proces...  
1      Repository: ILO-STATISTICS - Micro data proces...  
2      Repository: ILO-STATISTICS - Micro data proces...  
3      Repository: ILO-STATISTICS - Micro data proces...  
4      Repository: ILO-STATISTICS - Micro data proces...  
...                                                  ...  
32022  Repository: ILO-STATISTICS - Micro data proces...  
32023  Reposi

 We have a total of 32027 rows and 11 colums

#Evaluate the number of empty cells in each column and sum them

In [44]:
a.isnull().sum()

ref_area.label              0
indicator.label             0
source.label                0
sex.label                   0
classif1.label              0
time                        0
obs_value                3643
obs_status.label        22585
note_classif.label      30779
note_indicator.label    29267
note_source.label           0
dtype: int64

some of the columns obtain empty cells

#Now we check the number of both rows and columns

In [11]:
a.shape

(32027, 11)

#now lets fill the number of empty cells with zero's

In [12]:
a.fillna('0')

ref_area.label                                    indicator.label  \
0        Afghanistan  Employment outside the formal sector by sex an...   
1        Afghanistan  Employment outside the formal sector by sex an...   
2        Afghanistan  Employment outside the formal sector by sex an...   
3        Afghanistan  Employment outside the formal sector by sex an...   
4        Afghanistan  Employment outside the formal sector by sex an...   
...              ...                                                ...   
32022       Zimbabwe  Employment outside the formal sector by sex an...   
32023       Zimbabwe  Employment outside the formal sector by sex an...   
32024       Zimbabwe  Employment outside the formal sector by sex an...   
32025       Zimbabwe  Employment outside the formal sector by sex an...   
32026       Zimbabwe  Employment outside the formal sector by sex an...   

                    source.label    sex.label  \
0      LFS - Labour Force Survey   Sex: Total   
1      LFS - Labour Force Survey   Sex: Total   
2      LFS - Labour Force Survey   Sex: Total   
3      LFS - Labour Force Survey    Sex: Male   
4      LFS - Labour Force Survey    Sex: Male   
...                          ...          ...   
32022  LFS - Labour Force Survey  Sex: Female   
32023  LFS - Labour Force Survey  Sex: Female   
32024  LFS - Labour Force Survey  Sex: Female   
32025  LFS - Labour Force Survey  Sex: Female   
32026  LFS - Labour Force Survey  Sex: Female   

                                          classif1.label  time obs_value  \
0      Economic activity (Agriculture, Non-Agricultur...  2021  6059.359   
1      Economic activity (Agriculture, Non-Agricultur...  2021  1884.001   
2      Economic activity (Agriculture, Non-Agricultur...  2021  4175.358   
3      Economic activity (Agriculture, Non-Agricultur...  2021   4325.32   
4      Economic activity (Agriculture, Non-Agricultur...  2021  1237.423   
...                                                  ...   ...       ...   
32022  Economic activity (ISIC-Rev.4): R. Arts, enter...  2019         0   
32023  Economic activity (ISIC-Rev.4): S. Other servi...  2019    27.658   
32024  Economic activity (ISIC-Rev.4): T. Activities ...  2019   119.138   
32025  Economic activity (ISIC-Rev.4): U. Activities ...  2019         0   
32026  Economic activity (ISIC-Rev.4): X. Not elsewhe...  2019         0   

      obs_status.label note_classif.label  \
0                    0                  0   
1                    0                  0   
2                    0                  0   
3                    0                  0   
4                    0                  0   
...                ...                ...   
32022       Unreliable                  0   
32023  Break in series                  0   
32024  Break in series                  0   
32025       Unreliable                  0   
32026       Unreliable                  0   

                       note_indicator.label  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   
...                                     ...   
32022  Break in series: Methodology revised   
32023  Break in series: Methodology revised   
32024  Break in series: Methodology revised   
32025  Break in series: Methodology revised   
32026  Break in series: Methodology revised   

                                       note_source.label  
0      Repository: ILO-STATISTICS - Micro data proces...  
1      Repository: ILO-STATISTICS - Micro data proces...  
2      Repository: ILO-STATISTICS - Micro data proces...  
3      Repository: ILO-STATISTICS - Micro data proces...  
4      Repository: ILO-STATISTICS - Micro data proces...  
...                                                  ...  
32022  Repository: ILO-STATISTICS - Micro data proces...  
32023  Repository: ILO-ST

#lets check the number of countries from the data

In [46]:
a['ref_area.label'].nunique()

109

The total number of countries is found to be 109

In [69]:
fig = px.scatter(a, 
                 x='obs_value', 
                 y='sex.label',  
                 title='Gender behavior in each obs_value')
fig.update_traces(marker_size=4)
fig.show()

the graph of obs_values vs time

#lets look at the descriptive statistics of the column named obs_values

In [48]:
a['obs_value'].describe()

count     28384.000000
mean       1668.372981
std       11707.613403
min           0.000000
25%           6.057000
50%          42.616000
75%         345.274000
max      331409.227000
Name: obs_value, dtype: float64

On the stats we will focus on the average which is 1668.372981 and the standard deviation which is found to be 11707.613403 where by it is indicating the distance of each country from the average.

#Lets gather the data by indicator label and name it b

In [16]:
b=a.groupby(['indicator.label'])

#Check the first five rows from b

In [73]:
b.head(2)

ref_area.label                                    indicator.label  \
0    Afghanistan  Employment outside the formal sector by sex an...   
1    Afghanistan  Employment outside the formal sector by sex an...   

                source.label   sex.label  \
0  LFS - Labour Force Survey  Sex: Total   
1  LFS - Labour Force Survey  Sex: Total   

                                      classif1.label  time  obs_value  \
0  Economic activity (Agriculture, Non-Agricultur...  2021   6059.359   
1  Economic activity (Agriculture, Non-Agricultur...  2021   1884.001   

  obs_status.label note_classif.label note_indicator.label  \
0              NaN                NaN                  NaN   
1              NaN                NaN                  NaN   

                                   note_source.label  
0  Repository: ILO-STATISTICS - Micro data proces...  
1  Repository: ILO-STATISTICS - Micro data proces...

we can see several null values and we need to do something about them

In [49]:
b.fillna('0')

ref_area.label               source.label    sex.label  \
0        Afghanistan  LFS - Labour Force Survey   Sex: Total   
1        Afghanistan  LFS - Labour Force Survey   Sex: Total   
2        Afghanistan  LFS - Labour Force Survey   Sex: Total   
3        Afghanistan  LFS - Labour Force Survey    Sex: Male   
4        Afghanistan  LFS - Labour Force Survey    Sex: Male   
...              ...                        ...          ...   
32022       Zimbabwe  LFS - Labour Force Survey  Sex: Female   
32023       Zimbabwe  LFS - Labour Force Survey  Sex: Female   
32024       Zimbabwe  LFS - Labour Force Survey  Sex: Female   
32025       Zimbabwe  LFS - Labour Force Survey  Sex: Female   
32026       Zimbabwe  LFS - Labour Force Survey  Sex: Female   

                                          classif1.label  time obs_value  \
0      Economic activity (Agriculture, Non-Agricultur...  2021  6059.359   
1      Economic activity (Agriculture, Non-Agricultur...  2021  1884.001   
2      Economic activity (Agriculture, Non-Agricultur...  2021  4175.358   
3      Economic activity (Agriculture, Non-Agricultur...  2021   4325.32   
4      Economic activity (Agriculture, Non-Agricultur...  2021  1237.423   
...                                                  ...   ...       ...   
32022  Economic activity (ISIC-Rev.4): R. Arts, enter...  2019         0   
32023  Economic activity (ISIC-Rev.4): S. Other servi...  2019    27.658   
32024  Economic activity (ISIC-Rev.4): T. Activities ...  2019   119.138   
32025  Economic activity (ISIC-Rev.4): U. Activities ...  2019         0   
32026  Economic activity (ISIC-Rev.4): X. Not elsewhe...  2019         0   

      obs_status.label note_classif.label  \
0                    0                  0   
1                    0                  0   
2                    0                  0   
3                    0                  0   
4                    0                  0   
...                ...                ...   
32022       Unreliable                  0   
32023  Break in series                  0   
32024  Break in series                  0   
32025       Unreliable                  0   
32026       Unreliable                  0   

                       note_indicator.label  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   
...                                     ...   
32022  Break in series: Methodology revised   
32023  Break in series: Methodology revised   
32024  Break in series: Methodology revised   
32025  Break in series: Methodology revised   
32026  Break in series: Methodology revised   

                                       note_source.label  
0      Repository: ILO-STATISTICS - Micro data proces...  
1      Repository: ILO-STATISTICS - Micro data proces...  
2      Repository: ILO-STATISTICS - Micro data proces...  
3      Repository: ILO-STATISTICS - Micro data proces...  
4      Repository: ILO-STATISTICS - Micro data proces...  
...                                                  ...  
32022  Repository: ILO-STATISTICS - Micro data proces...  
32023  Repository: ILO-STATISTICS - Micro data proces...  
32024  Repository: ILO-STATISTICS - Micro data proces...  
32025  Repository: ILO-STATISTICS - Micro data proces...  
32026  Repository: ILO-STATISTICS - Micro data proces...  

[32027 rows x 10 columns]

Now our data seem to be clean therefore we can move forward with the analysis

#From b table we now check the maximum value to see if it correlates with descriptive stats from a column obs_value? in which country and sex.label

In [50]:
b.max()

<ipython-input-50-fe05da57cc16>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.max is deprecated. In a future version, a TypeError will be raised. Before calling .max, select only columns which should be valid for the function.
  b.max()


ref_area.label  \
indicator.label                                                     
Employment outside the formal sector by sex and...       Zimbabwe   

                                                              source.label  \
indicator.label                                                              
Employment outside the formal sector by sex and...  PC - Population Census   

                                                     sex.label  \
indicator.label                                                  
Employment outside the formal sector by sex and...  Sex: Total   

                                                                                       classif1.label  \
indicator.label                                                                                         
Employment outside the formal sector by sex and...  Economic activity (ISIC-Rev.4): X. Not elsewhe...   

                                                    time   obs_value  \
indicator.label                                                        
Employment outside the formal sector by sex and...  2022  331409.227   

                                                                                    note_source.label  
indicator.label                                                                                        
Employment outside the formal sector by sex and...  Repository: ILO-STATISTICS - Micro data proces...

The country is Zimbabwe and the value 331409.227 from the year 2022 which correlates

#Lets check for the middle value and see when was it obtained 

In [42]:
b.median()

<ipython-input-42-4035fb3efbb5>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.median is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  b.median()


time  obs_value
indicator.label                                                      
Employment outside the formal sector by sex and...  2019.0     42.616

The middle value also correlates to 42.616 and it was found in 2019

#Now we check the minimum value and country that its found from and sex.label

In [33]:
b.min()

<ipython-input-33-246853d5cbe6>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.min is deprecated. In a future version, a TypeError will be raised. Before calling .min, select only columns which should be valid for the function.
  b.min()


ref_area.label  \
indicator.label                                                     
Employment outside the formal sector by sex and...    Afghanistan   

                                                                                         source.label  \
indicator.label                                                                                         
Employment outside the formal sector by sex and...  HIES - EU Statistics on Income and Living Cond...   

                                                      sex.label  \
indicator.label                                                   
Employment outside the formal sector by sex and...  Sex: Female   

                                                                                classif1.label  \
indicator.label                                                                                  
Employment outside the formal sector by sex and...  Economic activity (Aggregate): Agriculture   

                                                    time  obs_value  \
indicator.label                                                       
Employment outside the formal sector by sex and...  2018        0.0   

                                                                                    note_source.label  
indicator.label                                                                                        
Employment outside the formal sector by sex and...  Repository: ILO-STATISTICS - Micro data proces...

The minumum value is 0.0 from the year 2018 which implies that the descriptive stats obtained above is correct

In [60]:
fig1 = px.histogram(a, 
                   x='ref_area.label',
                   y='obs_value', 
                   marginal='box', 
                   color_discrete_sequence=['blue'], 
                   title='Countries vs obs_values')
fig1.update_layout(bargap=0.1)
fig1.show()

#To see the behavior of Gender let us group the data from a-table by sex.label and create a new table named c

In [36]:
c=a.groupby(['sex.label'])

#The descriptive stats  will provide us with the exact values of gender behavior

In [39]:
c.describe()

time                                                         \
               count         mean       std     min     25%     50%     75%   
sex.label                                                                     
Sex: Female  10640.0  2019.508365  1.217539  2018.0  2018.0  2019.0  2021.0   
Sex: Male    10652.0  2019.508637  1.216966  2018.0  2018.0  2019.0  2021.0   
Sex: Other      83.0  2019.506024  1.108432  2018.0  2019.0  2020.0  2020.0   
Sex: Total   10652.0  2019.508637  1.216966  2018.0  2018.0  2019.0  2021.0   

                    obs_value                                              \
                max     count         mean           std    min       25%   
sex.label                                                                   
Sex: Female  2022.0    8717.0   924.754817   4629.065083  0.000   4.29800   
Sex: Male    2022.0    9594.0  1627.047159  11604.833013  0.000   6.06125   
Sex: Other   2021.0      40.0    40.591650     26.342579  9.121  20.02200   
Sex: Total   2022.0   10033.0  2360.460164  15473.585421  0.005   8.08700   

                                            
                50%        75%         max  
sex.label                                   
Sex: Female  31.539  264.67700   90251.158  
Sex: Male    39.800  309.85125  241138.046  
Sex: Other   36.028   69.30200      86.456  
Sex: Total   62.268  514.28400  331409.227

#Lets see a country that had the lowest obs_value

In [40]:
c.min()

<ipython-input-40-e9b1f7a680c2>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.min is deprecated. In a future version, a TypeError will be raised. Before calling .min, select only columns which should be valid for the function.
  c.min()


ref_area.label                                    indicator.label  \
sex.label                                                                       
Sex: Female    Afghanistan  Employment outside the formal sector by sex an...   
Sex: Male      Afghanistan  Employment outside the formal sector by sex an...   
Sex: Other           India  Employment outside the formal sector by sex an...   
Sex: Total     Afghanistan  Employment outside the formal sector by sex an...   

                                                  source.label  \
sex.label                                                        
Sex: Female  HIES - EU Statistics on Income and Living Cond...   
Sex: Male    HIES - EU Statistics on Income and Living Cond...   
Sex: Other                  LFS - Periodic Labour Force Survey   
Sex: Total   HIES - EU Statistics on Income and Living Cond...   

                                         classif1.label  time  obs_value  \
sex.label                                                                  
Sex: Female  Economic activity (Aggregate): Agriculture  2018      0.000   
Sex: Male    Economic activity (Aggregate): Agriculture  2018      0.000   
Sex: Other   Economic activity (Aggregate): Agriculture  2018      9.121   
Sex: Total   Economic activity (Aggregate): Agriculture  2018      0.005   

                                             note_source.label  
sex.label                                                       
Sex: Female  Repository: ILO-STATISTICS - Micro data proces...  
Sex: Male    Repository: ILO-STATISTICS - Micro data proces...  
Sex: Other   Repository: ILO-STATISTICS - Micro data proces...  
Sex: Total   Repository: ILO-STATISTICS - Micro data proces...

Both male and female from afghanistan were found to be 0.000 in the year 2018

#Lets see which country has the highest obs_value and when

In [51]:
c.max()

<ipython-input-51-26cf793d3be8>:1: FutureWarning: Dropping invalid columns in DataFrameGroupBy.max is deprecated. In a future version, a TypeError will be raised. Before calling .max, select only columns which should be valid for the function.
  c.max()


ref_area.label                                    indicator.label  \
sex.label                                                                       
Sex: Female       Zimbabwe  Employment outside the formal sector by sex an...   
Sex: Male         Zimbabwe  Employment outside the formal sector by sex an...   
Sex: Other           India  Employment outside the formal sector by sex an...   
Sex: Total        Zimbabwe  Employment outside the formal sector by sex an...   

                                   source.label  \
sex.label                                         
Sex: Female              PC - Population Census   
Sex: Male                PC - Population Census   
Sex: Other   LFS - Periodic Labour Force Survey   
Sex: Total               PC - Population Census   

                                                classif1.label  time  \
sex.label                                                              
Sex: Female  Economic activity (ISIC-Rev.4): X. Not elsewhe...  2022   
Sex: Male    Economic activity (ISIC-Rev.4): X. Not elsewhe...  2022   
Sex: Other               Economic activity (ISIC-Rev.4): Total  2021   
Sex: Total   Economic activity (ISIC-Rev.4): X. Not elsewhe...  2022   

              obs_value                                  note_source.label  
sex.label                                                                   
Sex: Female   90251.158  Repository: ILO-STATISTICS - Micro data proces...  
Sex: Male    241138.046  Repository: ILO-STATISTICS - Micro data proces...  
Sex: Other       86.456  Repository: ILO-STATISTICS - Micro data proces...  
Sex: Total   331409.227  Repository: ILO-STATISTICS - Micro data proces...

The maximum value is from Zimbabwe and found to be 331409.227 

#Conclusion

In conclusion Zimbabwe is found to have most of its employment outside the formal sector by both males and females while afghanistan is the least with both males and females. From the descriptive statistics the distance from each country to the average of the combined countries is high. Lastly Both the descriptive stats and the groupby method lay the same results, whereby for simplicity one can use the describe method.